File: acg-android/app/src/main/java/com/aicraftspeopleguild/acg/data/ApiClient.kt
package com.aicraftspeopleguild.acg.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException
import java.util.concurrent.TimeUnit

/** Thin OkHttp wrapper around the public L4 API. No auth required. */
class ApiClient(base: String = "https://aicraftspeopleguild.github.io") {
    private val baseUrl: String = base.trimEnd('/') + "/guild/Enterprise/L4"
    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()
    private val json = Json { ignoreUnknownKeys = true }

    @Throws(IOException::class)
    fun health(): Health = json.decodeFromString(getString("$baseUrl/api/health.json"))

    /** /runtime/tags.json — free-form; return the raw object so callers can drill in. */
    @Throws(IOException::class)
    fun runtimeTags(): JsonObject = json.parseToJsonElement(getString("$baseUrl/runtime/tags.json")).jsonObject

    data class EnterpriseSnapshot(
        val papers: Int, val members: Int, val programs: Int,
        val runs: Int, val tagEdges: Int, val authoredLinks: Int,
    )

    @Throws(IOException::class)
    fun enterprise(): EnterpriseSnapshot {
        val ent = runtimeTags()["enterprise"]?.jsonObject
            ?: return EnterpriseSnapshot(0, 0, 0, 0, 0, 0)
        // Fields come as either {value, quality, type} UDT wrappers or bare
        // primitives depending on the source. Handle both.
        fun pick(k: String): Int {
            val el: JsonElement = ent[k] ?: return 0
            val primitive: JsonPrimitive = when (el) {
                is JsonPrimitive -> el
                is JsonObject    -> (el["value"] as? JsonPrimitive) ?: return 0
                else             -> return 0
            }
            return primitive.content.toIntOrNull() ?: 0
        }
        return EnterpriseSnapshot(
            papers        = pick("paperCount"),
            members       = pick("memberCount"),
            programs      = pick("programCount"),
            runs          = pick("runCount"),
            tagEdges      = pick("tagEdges"),
            authoredLinks = pick("authoredLinks"),
        )
    }

    private fun getString(url: String): String {
        val req = Request.Builder().url(url).header("Accept", "application/json").build()
        http.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) throw IOException("HTTP ${resp.code} $url")
            return resp.body?.string().orEmpty()
        }
    }
}

File: phone/acg-ios/ACG/Data/ApiClient.swift
import Foundation

/// OkHttp-equivalent. Thin URLSession wrapper around the public L4
/// endpoints on the teslasolar Pages host.
final class ApiClient {
    let baseUrl: String

    init(base: String = "https://teslasolar.github.io/aicraftspeopleguild.github.io") {
        self.baseUrl = base.trimmingCharacters(in: .init(charactersIn: "/")) + "/guild/Enterprise/L4"
    }

    struct Health: Decodable {
        var paperCount: Int?
        var memberCount: Int?
        var lastUpdated: String?
        var apiVersion: String?
    }

    struct EnterpriseSnapshot {
        var papers = 0, members = 0, programs = 0
        var runs = 0, tagEdges = 0, authoredLinks = 0
    }

    func health() async throws -> Health {
        try await getJson(path: "/api/health.json", as: Health.self)
    }

    func enterprise() async throws -> EnterpriseSnapshot {
        let root: [String: Any] = try await getJsonObject(path: "/runtime/tags.json")
        guard let ent = root["enterprise"] as? [String: Any] else { return .init() }
        func pick(_ k: String) -> Int {
            if let n = ent[k] as? Int { return n }
            if let s = ent[k] as? String, let n = Int(s) { return n }
            if let d = ent[k] as? [String: Any] {
                if let n = d["value"] as? Int { return n }
                if let s = d["value"] as? String, let n = Int(s) { return n }
            }
            return 0
        }
        return .init(
            papers: pick("paperCount"), members: pick("memberCount"),
            programs: pick("programCount"), runs: pick("runCount"),
            tagEdges: pick("tagEdges"), authoredLinks: pick("authoredLinks"),
        )
    }

    // MARK: - helpers
    private func getJson<T: Decodable>(path: String, as type: T.Type) async throws -> T {
        let data = try await getData(path: path)
        return try JSONDecoder().decode(T.self, from: data)
    }
    private func getJsonObject(path: String) async throws -> [String: Any] {
        let data = try await getData(path: path)
        return (try JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
    }
    private func getData(path: String) async throws -> Data {
        guard let url = URL(string: baseUrl + path) else { throw URLError(.badURL) }
        let (data, resp) = try await URLSession.shared.data(from: url)
        if let http = resp as? HTTPURLResponse, http.statusCode != 200 {
            throw NSError(domain: "ACG.Api", code: http.statusCode,
                          userInfo: [NSLocalizedDescriptionKey: "HTTP \(http.statusCode) \(url)"])
        }
        return data
    }
}
